add first config test frame

lewismc opened the pull request:
@mjn0898 I'm creating a pull request for you.
I'm a bit concerned that this is the result of the entire projects work up until now... the rest of the work has been trivial. We're talking <10 lines of code for around 6 weeks' of work...
Moving forward, let's think about how you can make a contribution before the end of the project such that you can get a decent grade.
Please update this pull request as early as possible.
I also asked for every team member to be testing each others' code/pull requests. You can make a contribution by doing that but I have not seen anything like that from you yet. Please make sure you are testing your team members code and providing feedback.
Thanks

mjn0898 committed:
add first config test frame

mjn0898 committed:
three different config options with load in memory testing

mjn0898 commented:
I added a couple more tests now. 
To address your concerns a bit further: For all the previous work, which could be called, as you say, trivial, that I did, I had to mostly scrounge it up myself without receiving a direct task to work on. This made it difficult for me to make contributions that might be more meaningful since I felt I had no choice but to work on docs or the like, given there was nothing apparent in the implementation I could help with.
Now I'm sitting here trying to write tests for a config file that doesn't exist yet, so I'm making some assumptions about the format and then testing for the same things we already are testing, which is why I'm calling into previous test functions we already have, so I'm not sure measuring in lines of code is an appropriate metric. 
As for writing new tests, it doesn't seem obvious what functionality to focus on if we indeed want novel tests, and I didn't receive much feedback when mentioning this last week. 
I don't want to write a huge block of text, but I didn't want to ignore your disappointed message.

As for testing other member's code, I have been pulling other branches to try. Let me know if there is something you would like me to show you. In terms of contributions moving forward (besides working on the presentation), I am open to any suggestions.

I hope this clears some things up.

lewismc commented:
hi @mjn0898 thanks for writing some of your thoughts. I appreciate it. Please see my replies

> For all the previous work, which could be called, as you say, trivial, that I did, I had to mostly scrounge it up myself without receiving a direct task to work on. 

At the beginning everyone was given flexibility to work on what they were interested in as long as it landed under the larger umbrella task. Please see the original requirements document if you are unclear what that was. If you were unclear at any time, you should have reached out to me. I would highly encourage you to do this in your future tasks and assignments. There is nothing wrong with needing clarification on tasks. I do it literally every day.

> This made it difficult for me to make contributions that might be more meaningful since I felt I had no choice but to work on docs or the like, given there was nothing apparent in the implementation I could help with.

I was never aware that '[you] felt [you] had no choice'. If this is the situation you found yourself, and if that was in part due to my unclear communication/direction then I must apologize. There is absolutely loads within the implementation which you could work on... for example Dask. There are 18 open issues for you to choose to assist or take a lead on. 

> Now I'm sitting here trying to write tests for a config file that doesn't exist yet, so I'm making some assumptions about the format and then testing for the same things we already are testing, which is why I'm calling into previous test functions we already have

This is called test Driven Development. It's common practice in industry and results in some of the best codebases I've seen. I understand that you may or may not be comfortable writing tests for code which does not exist... but this is a task. If you need guidance or clarification on the task, then please ask me. The issue of working on tests was initially discussed a month or so ago IIRC. We can look at the notes, this is why I like to keep notes for the meetings. 

> so I'm not sure measuring in lines of code is an appropriate metric.

I agree that SLOC is not always a good metric for determining development progress... but in this case I brought it up because I thought it would prompt you to step up and take on the task at hand. I am not criticizing the number of lines of code. I'm also not criticizing how it is written. I was indicating my concern that this does not seem like enough work to cover month' worth of work. I hope this is clear.

> As for writing new tests, it doesn't seem obvious what functionality to focus on if we indeed want novel tests, 

You missed a few meetings when this was discussed. I understand you had examinations (which are extremely important) so I am not blaming you for anything. This is why I asked you to write ticket(s) for the work you planned to do as soon as possible so I could assist you with adequately scoping the work. You created [ISSUE-185](https://github.com/capstone-coal/pycoal/issues/185) 11 days ago creating a branch and subsequently updated it 2 days ago. So 2 weeks have gone by and there is not too much to show for that period of time. 
Again, maybe me asking more questions on the issue directly would have helped but I did contact you and the team on Slack several times during that course of time. 

> and I didn't receive much feedback when mentioning this last week.

IIRC you dropped off the call early as you had an appointment so you didn't mention it. Maybe I am wrong here. 

>  I don't want to write a huge block of text, but I didn't want to ignore your disappointed message.

Let me be clear _**I am not disappointed**_. The reason I stated "Moving forward, let's think about how you can make a contribution before the end of the project such that you can get a decent grade." is because I am open to you making contributions. So let's just focus on that.

> As for testing other member's code, I have been pulling other branches to try.

On several occasions I've asked for the entire team to review each others' code and provide feedback. This is immensely important. I have yet to see you provide any feedback. I'll repeat what I said above... "You can make a contribution by doing that but I have not seen anything like that from you yet. Please make sure you are testing your team members code and providing feedback."

I hope the above clarifies my position. Let me conclude

* if you are unsure, then ask me
* provide feedback on your team members' code and I hope that they will do the same for you
* my goal is to ensure that everyone has a fair crack at the whip and that everyone has the opportunity to obtain a good grade. Working remotely like this is not a perfect science. Things can and do go wrong. Communication is key. Please make sure to keep me updated if you encounter issues. I will assist you. Believe me.

mjn0898 commented:
Understood, thanks for your clarifications. I would appreciate suggestions on what to focus on for the next two weeks, as perhaps you have a better idea of what is most critical to the project. In the meantime, I shall continue expanding the test suite as ideas come to me and of course validating current pull requests and updating the relevant member of the team with feedback.

lewismc commented on pycoal/tests/mineral_test.py:
```diff
@@ -81,7 +82,36 @@ def test_classify_image():
         # verify that every pixel has the same classification
         assert numpy.array_equal(expected.asarray(), actual.asarray())
 
+# verify that classified images have valid classifications when using config file
```
This is looking better. Please go ahead and implement the **config.ini** file and then work on the configuration management logic. This will involve 1) loading the file and also 2) augmenting the various [__init__](https://github.com/capstone-coal/pycoal/blob/master/pycoal/mineral.py#L288) functions in mineral, mining and environment.py to accommodate config.ini.

mjn0898 replied:
I've updated the mineral.py file to support setting the algorithm (SAM or avngDNN) and implementation via the config file. Now all that is needed is to add the different parallelized algorithms as functions to call into.

lewismc commented on pycoal/tests/mineral_test.py:
```diff
@@ -81,7 +82,36 @@ def test_classify_image():
+def test_classify_image_dask():
+    # set our config file to dask
+    config = configparser.ConfigParser()
+    config.read('config.py')
```
**config.ini**

mjn0898 replied:
fixed

lewismc commented on pycoal/tests/mineral_test.py:
```diff
@@ -81,7 +82,36 @@ def test_classify_image():
+    
+@with_setup(None, _test_classify_image_teardown)
+def test_classify_image_pytorch():
+    # set our config file to pytorch
```
Your not setting the config file, you are setting the `[processing][impl]` config key.

mjn0898 replied:
updated

mjn0898 committed:
update comment setting ['processing']['impl']

mjn0898 committed:
move config file parsing to setup function

mjn0898 committed:
Create config.ini

mjn0898 committed:
Update config.ini

mjn0898 committed:
Update mineral_test.py

mjn0898 committed:
Update config.ini

mjn0898 committed:
Update config.ini

mjn0898 committed:
Update config.ini

mjn0898 committed:
Update config.ini

mjn0898 committed:
Update mineral.py

Initial changes to support setting algorithm from config file

mjn0898 committed:
Update mineral.py

choose algorithm and implementation based on config file

lewismc commented on pycoal/mineral.py:
```diff
@@ -285,7 +286,7 @@ def avngDNN(image_file_name, classified_file_name, model_file_name,
 class MineralClassification:
 
-    def __init__(self, algorithm=SAM, **kwargs):
+    def __init__(self, '''algorithm=SAM,''' **kwargs):
```
Why wrap in single quotes. If you are not using then simply remove. 
I do do think we want to retain this parameter however or else we introduce an API regression. Please remove the quotes.

lewismc commented on pycoal/mineral.py:
```diff
@@ -303,9 +304,35 @@ def __init__(self, algorithm=SAM, **kwargs):
+        set_impl = config['processing']['impl']
+        
+        if set_algo == 'SAM':
+                if set_impl == 'serial':
```
By the looks of it the formatting/indents is off here. Please make sure you follow the existing formatting e.g. 4 space indents.

lewismc commented on pycoal/mineral.py:
```diff
@@ -303,9 +304,35 @@ def __init__(self, algorithm=SAM, **kwargs):
+        if set_algo == 'SAM':
+                if set_impl == 'serial':
+                        self.algorithm = SAM
+                        '''
```
Never leave code which is not being used. Please uncomment this logic.

lewismc replied:
Also, eventually, I want the default to by pytorch. Simply because it is most performant.

lewismc commented on pycoal/mineral.py:
```diff
@@ -303,9 +304,35 @@ def __init__(self, algorithm=SAM, **kwargs):
+                        self.algorithm = SAM_joblib
+                        '''
+        elif set_algo == 'avngDNN':
+                if set_impl == 'serial':
```
This is unnecessary duplicated code. Please refactor it and again remove commented-out code.

lewismc commented on pycoal/tests/config.ini:
```diff
@@ -0,0 +1,3 @@
+[processing]
```
Add license header like every other file.

lewismc commented on pycoal/tests/mineral_test.py:
```diff
@@ -81,7 +87,34 @@ def test_classify_image():
+    test_classify_image()
+    test_classify_image_in_memory()
+
+@with_setup(_init_config_parser, _test_classify_image_teardown)
```
We need to make sure that these tests are running on small images. If this is the case then please resolve this conversation. Thanks

lewismc commented on pycoal/tests/mineral_test.py:
```diff
@@ -81,7 +87,34 @@ def test_classify_image():
+    test_classify_image()
+    test_classify_image_in_memory()
+
+    
```
You should also introduce a test which tests for incorrect/wrong/dummy configuration. This tests the new config-loading logic you've implemented.

lewismc reviewed (commented):
Please see my comments @mjn0898 
Still some work to do here.

mjn0898 committed:
Update mineral.py

raise exceptions from config file and change setting of self.algorithm. default algo to SAM_pytorch

mjn0898 committed:
Added license header

mjn0898 committed:
Rename pycoal/tests/config.ini to pycoal/config.ini

mjn0898 committed:
Update mineral_test.py

lewismc commented on pycoal/mineral.py:
```diff
@@ -303,9 +305,33 @@ def __init__(self, algorithm=SAM, **kwargs):
+        except IOError:
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), 'config.ini')
+
+        set_algo = None
```
These should not be `None` they should be set to the default values.

mjn0898 replied:
The algorithm is set as default above. This being set to None allows me not to update self. algorithm, leaving it as default. The set_algo is just for the value pulled from the config file

mjn0898 committed:
Update mineral.py

mjn0898 committed:
Update mineral.py

mjn0898 committed:
Update mineral.py

lewismc commented:
@mjn0898 please rebase against master branch.

mjn0898 committed:
Merge branch 'master' into ISSUE-185

mjn0898 commented:
I have resolved the conflict with master branch

lewismc commented:
Just merged into master. My feeling is that there is more work to be done here.

lewismc commented:
Thanks for the work @mjn0898

mjn0898 commented:
Sure. There is still some more work to be done in the testing files, as well as looking into extending the config file. However, I can make a separate branch for that so that we can use the current changes.
